chain/chain/src/chain.rs

```diff
@@ -1346,6 +1346,13 @@ impl Chain {
         shard_id: ShardId,
         sync_hash: CryptoHash,
     ) -> Result<ShardStateSyncResponseHeader, Error> {
+        // Check cache
+        let key = StateHeaderKey(shard_id, sync_hash).try_to_vec()?;
+        match self.store.owned_store().get_ser(ColStateHeaders, &key) {
+            Ok(Some(header)) => return Ok(header),
+            _ => {}
+        }
```

Comment:
This feels suboptimal. We should have cache in between and not query database directly. Also I believe you can use `.store()` instead of `.owned_store()`

Reply:
We shouldn't optimize everything just because we can. In comparison to other operations and how nodes operate, I see no win by having additional layer of caching at this point.

++ for .store()

Reply:
I think this operation can happen quite frequently, so caching might help.

Reply:
You at least will need to get then all State Parts which are not cached. And there may be thousands of State Parts, thousands of State Part Requests **per one** State Header Request.
And it's still a droplet in the ocean in comparison to Block queries which commits to the database every few seconds / State Header requests appear once a new fresh node connected or epoch is changed.

Reply:
Oh I didn't realize this is only headers.